### PR TITLE
Innodb migrate

### LINKED ISSUE #4993: Use InnoDB instead of MyIsam (Trac #4993)
_Original ticket http://trac.elgg.org/ticket/4993 on 43059497-04-27 by trac user coldtrick, assigned to trac user coldtrick._

Elgg version: Github Master

Elgg should use innodb instead of the myisam storage engine.

Innodb is in MySQL (5.5) the new standard storage engine. It provides higher concurrency use, provides transactions and foreign key support.

Currently it is not possible to migrate Elgg to innodb because Elgg uses some full text indexes (that do not exist in InnoDB). But those indexes don't work as expected (they are slower compared to normal 'LIKE' queries, therefore it is no problem to drop those indexes and move to InnoDB.

This will require changes in:
- Search plugin: remove use of fulltext queries
- Core: 
  - schema sql to use innodb + removal of full text indexes
  - upgrade script to upgrade database engine to innodb and drop full text indexes
- deprecate min_search_length (config) with default to 9999? or should we drop this?

### PR DESCRIPTION
**Pull request by jdalsem**
chore(db): changed db storage engine to InnoDB

fixes #4993

BREAKING CHANGE:
To be able to still provide support for MySQL 5.5 combined with InnoDB
the FULLTEXT indices have been dropped. This effects how search works
internally.

Mysql info about transferring from MyISAM to InnoDB: http://dev.mysql.com/doc/refman/5.7/en/converting-tables-to-innodb.html

### CODE REVIEW

**Review by hypeJunction — COMMENTED** [review]
Are there any benchmarks? Should probably be a feature to appear in logs.

**Comment by jdalsem** [comment]
benchmarks on the storage engine change?

**Comment by jdalsem** [comment]
i actually do not see this as a feature (as Elgg doesn't provide any feature)... plus it could have been done years ago... it is more the preference of the DBA at this moment

**Comment by hypeJunction** [comment]
Benchmarks on using partial matches in search

**Comment by mrclay** [comment]
Awesome. Can someone point me to a summary of how search behavior will change?

**Comment by jdalsem** [comment]
The following changes to the way search behaves are:

_Assuming following search query:_
Query A: Search for "Jeroen alsem" (Missing D in last name)
Query B: Search for "alsem" (Missing D in last name)
A user exists with name "Jeroen Dalsem"

**Currently using FULLTEXT indexes with MATCH query:**
QUERY A: Will perform a MATCH on both Jeroen and alsem... will hit on Jeroen and return the user as a result.
QUERY B: No MATCH hit for "alsem"... no results

**New way A (single LIKE):**
Query A: No results as there is no result for "user.name LIKE %Jeroen alsem%"
Query B: Returns the user as it matches for the literal "user.name LIKE %alsem%"

**New way B (multiple LIKES):**
Query A: Returns the user because "user.name LIKE '%Jeroen%' AND user.name LIKE '%alsem%'"
Query B: Returns the user as it matches for the literal "user.name LIKE %alsem%"

Benefits of A and B are partial matches will also match. Downside of the New Way A is that is requires a full sentence match. That is why New Way B splits the query into multiple words to match.

**Benchmarks**
Created 5000 blogs with the hypeFaker. Used apache benchmark to query (as a logged out user) only the blog entities. Tested only the difference between LIKE vs FULL TEXT MATCH, so no InnoDB changes.
Tested 50 sequential requests and 50 requests by 5 users concurrent.

**Results**
_Single word query, 1:50:_
FT MATCH: 1614 ms
          LIKE: 1585 ms (-29)

_Single word query, 5:50:_
FT MATCH: 2024 ms
          LIKE: 1964 ms (-60)

_2 word query, 1:50:_
FT MATCH: 1400 ms
    LIKE (A): 1369 ms (-31)

_2 word query, 5:50:_
FT MATCH: 1787 ms
    LIKE (A): 1672 ms (-115)
    LIKE (B): 1754 ms (-33)

Based on these numbers we can say that using LIKE queries vs using FULL TEXT indices for search does not make performance worse (even makes it slightly better). Search results will be different however (not worse or better, just different). Currently implemented in this PR is New Way A.. i will update it to be New Way B.

File: engine/lib/upgrades/2016102600-3.0.0-db-engine-to-innodb-d636305e6aa305a5.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<?php
+/**
+ * Elgg 3.0.0 upgrade 2016102600
+ * db-engine-to-innodb
+ *
+ * Converts table engine to innodb
+ */
+
+set_time_limit(0);
```
Comment (mrclay) [inline engine/lib/upgrades/2016102600-3.0.0-db-engine-to-innodb-d636305e6aa305a5.php]: Any ideas on how long this conversion takes?

Comment (jdalsem) [inline engine/lib/upgrades/2016102600-3.0.0-db-engine-to-innodb-d636305e6aa305a5.php]: could take some time (tens of minutes) on big databases, but depends on table size and CPU power of DB server

**Pull request merged**